File: 06_Searching/src/Fair_Candy_Swap_888.java
import java.util.Arrays;

class Fair_Candy_Swap_888 {
    public int[] fairCandySwap(int[] aliceSizes, int[] bobSizes) {
        Arrays.sort(bobSizes);

        int sumAlice = 0, sumBob = 0;

        for (int sum : aliceSizes) {
            sumAlice += sum;
        }
        for (int sum : bobSizes) {
            sumBob += sum;
        }

        // Corrected: difference between Bob and Alice's sums, no abs
        int diff = (sumBob - sumAlice) / 2;

        int[] ans = new int[2];

        for (int i = 0; i < aliceSizes.length; i++) {
            int start = 0;
            int end = bobSizes.length - 1;

            while (start <= end) {
                int mid = start + (end - start) / 2;

                // Corrected: check aliceSizes[i] + diff
                if (bobSizes[mid] == aliceSizes[i] + diff) {
                    ans[0] = aliceSizes[i];
                    ans[1] = bobSizes[mid];
                    return ans;
                } else if (bobSizes[mid] > aliceSizes[i] + diff) {
                    end = mid - 1;
                } else {
                    start = mid + 1;
                }
            }
        }

        return ans;
    }
}

File: 06_Searching/src/Fair_Candy_Swap_888.js
/**
 * @param {number[]} aliceSizes
 * @param {number[]} bobSizes
 * @return {number[]}
 */
const fairCandySwap = function(aliceSizes, bobSizes) {
    let aliceSum = aliceSizes.reduce((accu,currentValue) => accu+currentValue);
    let bobSum = bobSizes.reduce((accu,currentValue) => accu+currentValue);

    bobSizes.sort((a,b) => a-b);

    let diff = (bobSum - aliceSum) /2;

    for (let i = 0; i < aliceSizes.length; i++) {
        let start = 0;
        let end = bobSizes.length-1;

        while (start <= end) {

            let mid = start + ((end-start) >> 1);
            let shouldAliceCandy = aliceSizes[i] + diff;

            if (shouldAliceCandy === bobSizes[mid]) {
                return [aliceSizes[i],bobSizes[mid]];
            } else if (shouldAliceCandy < bobSizes[mid]) {
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
    }

    return [-1,-1];
};

console.log(fairCandySwap([1,1],[2,2]),
fairCandySwap([1,2],[2,3]),
fairCandySwap([2],[1,3]))